src/ALCProxy.Communication/IProxyServer.cs

```diff
@@ -0,0 +1,157 @@
+        {
+            currentLoadContext = AssemblyLoadContext.GetLoadContext(Assembly.GetExecutingAssembly());
+            if (genericTypes != null)
+            {
+                instanceType = instanceType.MakeGenericType(genericTypes.Select(x => ConvertType(x)).ToArray());
+            }
+            constArgTypes = constArgTypes.Select(x => ConvertType(x)).ToList();
+            var constructorParams = DeserializeParameters(serializedConstParams, constArgTypes);
+            SetInstance(instanceType, constArgTypes.ToArray(), constructorParams);
+        }
+        /// <summary>
+        /// Create the instance of the object we want to proxy
+        /// </summary>
+        /// <param name="instanceType">the type of the object we want</param>
+        /// <param name="constructorTypes">The list of types that the constructor of the object takes in as an argument</param>
+        /// <param name="constructorArgs">The physical objects that are the parameters to the constructor</param>
+        protected void SetInstance(Type instanceType, Type[] constructorTypes, object[] constructorArgs)
+        {
+            var ci = instanceType.GetConstructor(constructorTypes);
+            instance = ci.Invoke(constructorArgs);
+        }
+        /// <summary>
+        /// Takes a Type that's been passed from the user ALC, and loads it into the current ALC for use. 
+        /// TODO: Figure out if we actually need to load here, or we can search the ALC for our loaded dependency which may already be loaded.
+        /// </summary>
+        protected Type ConvertType(Type toConvert)
+        {
+            string assemblyPath = Assembly.GetAssembly(toConvert).CodeBase.Substring(8);
+            if (toConvert.IsPrimitive || assemblyPath.Contains("System.Private.CoreLib")) //Can't load/dont want to load extra types from System.Private.CoreLib
+            {
+                return toConvert;
+            }
+            //TODO can we look for the existence of the loaded assembly here, instead of just doing the load?
+            return currentLoadContext.LoadFromAssemblyPath(assemblyPath).GetType(toConvert.FullName);
+        }
+        public object CallObject(MethodInfo targetMethod, IList<object> streams, IList<Type> argTypes)
+        {
+            //Turn the memstreams into their respective objects
+            argTypes = argTypes.Select(x => ConvertType(x)).ToList();
+            object[] args = DeserializeParameters(streams, argTypes);
+            MethodInfo[] methods = instance.GetType().GetMethods();
+            MethodInfo m = FindMethod(methods, targetMethod, argTypes.ToArray());
+            if (m.ContainsGenericParameters)
+            {
+                //While this may work without the conversion, we want it to uphold the type-load boundary, don't let the passed in method use anything from outside the target ALC
```

Comment:
I don't understand this requirement. I think it would be good to split the responsibilities here.
* The communication classes should not make a decision which assembly gets loaded where
* The ALC itself should make the final decisions where to load assemblies from
* The communication classes just need to make sure to load any types "by name" through the target ALC (so that it can make the right decisions).

Reply:
I'm a little confused on what you're trying to say here. Is this like the examples in the docs of using AssemblyLoadResolver to get the assemblies instead? The current implementation relies on making sure that any types that are required by the proxied object are loaded in from the proxy API, from the same assemblies that we're using for the default ALC where the proxy is created. I'm not sure how we can make sure everything will work without doing any direct loads from this API, but I may just be misunderstanding what you're suggesting...

Reply:
We need to load *through* the right ALC, but it doesn't matter where the actual assembly ends up being loaded to.
So in this case instead of calling LoadAssemblyByPath, simply call Load with a simple name, It's up to the ALC to decide how to load the assembly. That way you get to avoid the problem of not loading CoreLib twice (as ALCs will typically handle that correctly). It will also typically make sure that all framework types are shared, so for example if you would have an XElement as a parameter, right now you would try to load System.Private.Xml twice... which is generally not a good idea (it would probably work though).

The only visible change would be that the initial proxy creation would only specify assembly name (not a full path) and it would be up to the ALC to find the file. So you would probably need to sublass the custom ALC and implement some logic in Load - which most custom ALCs will do anyway.

Note that this will give you the easy "differences" - as each ALC may decide to load from different place. So for example you IPlugin may come from different places and thus have different version. But that is the desired behavior in the end.

Reply:
I've got the proxy system working now when you pass in an `AssemblyName`. Do you think it may be work to keep the file path option working to allow for people to use non-subclassed `AssemblyLoadContext`s for proxies, or will that just lead to more confusion?

Reply:
I would remove it - beside the potential confusion, I'm no aware of any other place where we explicitly load files into an ALC in a similar manner - so it's not a common patter either. I prefer simplicity in this case. If there's a need for something else we can extend it later on.